Add libc++ github actions workflow to replace buildkite

EricWF opened the pull request:
This change ports almost all of the linux buildkite builders to github actions.

I would like to have this transition occur as soon as possible.

boomanaiden154 commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,137 @@
+  cancel-in-progress: true
+
+permissions:
+  checks: write
```
Can we set the permissions to read only at the top level of the file and add additional permissions at the job level? That's what we're trying to do across the rest of the project to help reduce the possibility of mistakenly adding permissions where they aren't needed.

EricWF replied:
I think it's vestigial. Removing.

boomanaiden154 replied:
Uploading artifacts might need additional permissions. Not completely sure how that is supposed to work.

tstellar replied:
I don't think you need any permissions for uploading artifacts.  We have at least one workflow that does it with read-only permissions:  https://github.com/llvm/llvm-project/blob/main/.github/workflows/libclang-abi-tests.yml

EricWF replied:
All of the permissions have been removed and it seems to work great.

boomanaiden154 commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,137 @@
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 5
```
Is there a reason the fetch depth here (and below) is 5? As far as I understand, building the code should just need the top commit (although fetching 5 vs 1 is a negligible time impact in clone time).

EricWF replied:
This can go. I was attempting to do a thing with git diff, but that thing is long gone.

EricWF commented on libcxx/utils/ci/buildkite-pipeline-full.yml:
Yes, I just need to figure out how to use it.

We neet to use this for phabricator triggered CI runs, and the other one when triggered from github.

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+            "**/*.abilist"
+  stage2:
+    runs-on: libcxx-runners-8
+    needs: [ stage1  ]
```
Nit: double-space

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+    needs: [ stage1  ]
+    continue-on-error: false
+    strategy:
+      fail-fast: true
```
As discussed, we should not fail fast for jobs outside of stage1.

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+name: Build and Test libc++
```
I would like us to add a comment that explains a few aspects of this setup, because some of these choices are not arbitrary. Something like this:

```
# This file defines pre-commit CI for libc++ [bla bla bla].
#
# We split the configurations in multiple stages with the intent of saving compute time
# when a job fails early in the pipeline. This is why the jobs are marked as `continue-on-error: false`.
# We try to run the CI configurations with the most signal in the first stage.
# 
# TODO: Explain the fail-fast behavior (enabled for the first stage but not the other ones).
```

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+      fail-fast: true
+      matrix:
+        config: [ 'generic-cxx26', 'generic-cxx03', 'generic-modules' ]
+        cc: [ 'clang-18' ]
```
Instead of hardcoding `clang-18` here, we should define env variables like

```
env:
    # LLVM POST-BRANCH bump version
    # LLVM POST-BRANCH add compiler test for ToT - 1, e.g. "Clang 17"
    # LLVM RELEASE bump remove compiler ToT - 3, e.g. "Clang 15"
    LLVM_HEAD_VERSION: "18"   # Used compiler, update POST-BRANCH.
    GCC_STABLE_VERSION: "13"
```

and then we should use `clang-${LLVM_HEAD_VERSION}` or whatever makes the most sense with GH actions.

tstellar replied:
You can't reference environment variables at the point in the job definition.  One solution I've used is to add a 'variable' job at the very beginning of the workflow that collects all the global variables and outputs them for the other jobs to use.  We do this in the [release binaries workflow](https://github.com/llvm/llvm-project/blob/main/.github/workflows/release-binaries.yml#L23)

There is another option, that I haven't tried, but might work, which is to define repository-level [variables](https://docs.github.com/en/actions/learn-github-actions/variables).  The downside of this is that you would need an admin to change the settings although for something like LLVM_HEAD_VERSION, we may be able to add some automation that keeps it up-to-date at all times.

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+          CC: ${{ matrix.cc }}
+          CXX: ${{ matrix.cxx }}
+      - uses: actions/upload-artifact@v3
+        if: always()
```
```suggestion
        if: always() # Upload artifacts even if the test suite fails
```

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+      - name: ${{ matrix.config }}
+        run: libcxx/utils/ci/run-buildbot ${{ matrix.config }}
+        env:
+          CC: 'clang-18'
+          CXX: 'clang++-18'
```
```suggestion
          CC: ${{ config.cc }}
          CXX: ${{ config.cxx }}
```

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+      fail-fast: true
+      matrix:
+        config: [ 'generic-cxx11', 'generic-cxx14', 'generic-cxx17',
+                  'generic-cxx20', 'generic-cxx23' ]
```
Here we need to define `cc` and `cxx` to `clang-${LATEST}`.

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+      - name: ${{ matrix.config }}
+        run: libcxx/utils/ci/run-buildbot ${{ matrix.config }}
+        env:
+          CC: clang-18
+          CXX: clang++-18
```
```suggestion
          CC: clang-${{ LLVM_HEAD_VERSION }}
          CXX: clang++-${{ LLVM_HEAD_VERSION }}
```

EricWF replied:
I think I'm going to have to punt on these requested changes for the moment.

Github action doesn't support env substitution in the context we need it. So a more involved solution will be needed.

Rest assured I have plans to clean this up in short order.

ldionne commented on libcxx/utils/ci/buildkite-pipeline.yml:
```diff
@@ -101,291 +101,13 @@ environment_definitions:
-  env:
-    <<: *common_env
-    <<: *absolute_path_clang
-    ENABLE_STD_MODULES: 'On'
```
Those have been dropped in the added Github Actions, we should add them back to maintain the same coverage.

EricWF replied:
Done, except for std::modules, which previously hijacked the C++23 and C++26 build, making them not follow the "test as our users do" principle for basic use of C++23 and C++26. As a replacement, I added a stage 3 job that runs with std::modules in C++26

ldionne commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,112 @@
+name: Build and Test libc++
+
+on: pull_request
```
We need a path selector to only trigger on the following paths: `libcxx|libcxxabi|libunwind|runtimes|cmake`

EricWF committed:
Add libc++ github actionns workflow to replace buildkite.

This adds a github actions replacement for most of the buildkite
pipeline. There are  a lot of benefits to using actions over buildkite,
so I'll spare the reader reciting them all. The primary reasons are:

 * More flexability
 * Easier for other contributors to maintain.
 * Easier to scale

EricWF commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,180 @@
+# This file defines pre-commit CI for libc++ [bla bla bla].
```
lol I didn't even catch that bit. My bad.

EricWF commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,180 @@
+
+env:
+  CMAKE: "/opt/bin/cmake"
+  # LLVM POST-BRANCH bump version
```
I hate TODO's. I'm just going to do it. Not in this change, but none the less.

EricWF commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,180 @@
+      fail-fast: false
+      max-parallel: 8
+      matrix:
+        config: [ 'generic-no-threads', 'generic-no-filesystem', 'generic-no-random_device',
+          'generic-no-localization', 'generic-no-unicode', 'generic-no-wide-characters',
+          'generic-no-experimental', 'generic-no-exceptions', 'generic-abi-unstable',
+          'generic-hardening-mode-fast',
+          'generic-hardening-mode-fast-with-abi-breaks',
+          'generic-hardening-mode-extensive',
+          'generic-hardening-mode-debug', 'generic-with_llvm_unwinder', 'generic-static', 'generic-merged',
+          'generic-modules-lsv', 'generic-no-tzdb' ]
```
Using json here seems canonical. All the docs do it.

But  in the spirit of your comment, I've made the array one item per line, and I've sorted it.
I've done this for all the config lists.

EricWF commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,180 @@
+
+
+# Job breakdown:
+#   The jobs run in three stages to save resources while providing actionalable feedback as soon as possible.
```
I deleted the comment.

EricWF committed:
Address final review comments

mordante commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,196 @@
+        clang_tidy: [ 'ON' ]
+        include:
+          - config: 'generic-gcc'
+            cc: 'gcc-13'
```
Is it not possible to use the `GCC_STABLE_VERSION` here and `LLVM_HEAD_VERSION` in other places. I added these to buildkite to reduce the maintenance while updating compilers.

mordante replied:
Thanks for the info.

mordante commented on .github/workflows/libcxx-build-and-test.yaml:
```diff
@@ -0,0 +1,196 @@
+          'generic-with_llvm_unwinder'
+        ]
+        machine: [ 'libcxx-runners-8' ]
+        std_modules: [ 'OFF' ]
```
@EricWF It seems this turns off the testing of the modules in our CI. Why was this changed?

EricWF replied:
It turns off testing standard modules for the above configurations, yes.

Modules are still enabled for the `generic-cxx26` configuration below.

Turning on module for these configurations means that we no longer test these configurations without modules, correct?

mordante replied:
As discussed privately not entirely. Based on your suggestion to manually build modules instead of relying on CMake I've been working on a patch. The preparation patch for that removes this line. Since your suggested approach is a lot simpler and solves the concerns you voiced I think we can keep this as is for now and go for the simpler approach.
